### PR TITLE
Let ChangesetBuilder.invalidated search even container elements

### LINKED ISSUE #6221: When picking a Boolean value via the dropdown doesn't recalculate correctly.
https://user-images.githubusercontent.com/4699705/230432487-76f9bfb8-25c2-4083-98a8-7e4325729018.mp4

When I choose a Boolean value from a dropdown it doesn't appear to recalculate correctly.
When I then change the value it then recalculates correctly. And when I change back to the first value it calculates fine.

### PR DESCRIPTION
**Pull request by JaroslavTulach**
### Pull Request Description

Will fix #6221 by making sure IR element with UUID is found even it is not a leaf. 

### Checklist

Please ensure that the following checklist has been satisfied before submitting the PR:

- [x] All code follows the
      [Scala](https://github.com/enso-org/enso/blob/develop/docs/style-guide/scala.md),
      [Java](https://github.com/enso-org/enso/blob/develop/docs/style-guide/java.md),
      style guides. 
- All code has been tested:
  - [x] Unit tests have been written where possible.

### CODE REVIEW

File: engine/runtime/src/main/scala/org/enso/compiler/context/ChangesetBuilder.scala [hunk 1]
```diff
@@ -336,17 +336,40 @@ object ChangesetBuilder {
...
-          Node.fromIr(ir).foreach(acc.add)
+    def go(input: IR, acc: Tree): Unit = {
+      if (input.children.isEmpty) {
+        Node.fromIr(input).foreach(acc.add)
+      } else {
+        if (input.getExternalId.nonEmpty) {
+          val chTree = new Tree()
+          input.children.map(go(_, chTree))
+
+          def fillInHoles(p: Int, n: Node): Int = {
```
Comment (JaroslavTulach) [inline engine/runtime/src/main/scala/org/enso/compiler/context/ChangesetBuilder.scala]: Originally the algorithm searched only in leaf elements without any children. However those elements are not continuous - they may have holes among them. Moreover their parent elements may "cover those holes". This is what happens in the case of adding `Boolean.True` - there is a parent component (representing the whole expression) that has `getExternalId` and it should be found and reported as `invalidated` in my opinion.

File: engine/runtime/src/test/scala/org/enso/compiler/test/context/ChangesetBuilderTest.scala [hunk 2]
```diff
@@ -375,10 +375,78 @@ class ChangesetBuilderTest extends CompilerTest {
...
+      val atId   = "0d4b3071-3f7d-472a-bfa7-af118222894e"
+      val at     = findIR(ir, atId)
+      val atCode = findCode(code, at)
+      atCode shouldBe "text2.replace text1 text3 (Case_Sensitivity.Insensitive)"
+
+      val all = new ChangesetBuilder(Rope(code), ir).invalidated(Seq(edit))
+
+      all
+        .map(n => n.externalId.getOrElse(n.internalId))
+        .map(findCode(code, ir, _)) should contain theSameElementsAs Seq(
```
Comment (JaroslavTulach) [inline engine/runtime/src/test/scala/org/enso/compiler/test/context/ChangesetBuilderTest.scala:412]: This `findCode` call replaces UUIDs with their code, if such UUIDs are present in metadata (and thus in IR).

File: engine/runtime/src/test/scala/org/enso/compiler/test/context/ChangesetBuilderTest.scala [hunk 3]
```diff
@@ -375,10 +375,78 @@ class ChangesetBuilderTest extends CompilerTest {
...
+          |{"ide":{"node":{"ed3e40fb-d19e-4fb6-bebd-f74ee179ae08":{"position":{"vector":[52.000004,140.00055]},"intended_method":null,"uploading_file":null,"selected":false,"visualization":null},"0d4b3071-3f7d-472a-bfa7-af118222894e":{"position":{"vector":[-100.0,-40.0]},"intended_method":null,"uploading_file":null,"selected":false,"visualization":{"name":{"content":{"content":"JSON"}},"project":"Builtin"}},"54493640-4981-4437-ad6e-fc80ab1b581f":{"position":{"vector":[-111.0,140.00055]},"intended_method":null,"uploading_file":null,"selected":false,"visualization":null},"c143e355-9d64-4780-b387-dd364eefa2f9":{"position":{"vector":[9.723406,95.0]},"intended_method":null,"uploading_file":null,"selected":false,"visualization":null}},"import":{},"project":null}}
```
Comment (4e6) [inline engine/runtime/src/test/scala/org/enso/compiler/test/context/ChangesetBuilderTest.scala]: this line is unnecessary

File: engine/runtime/src/test/scala/org/enso/compiler/test/context/ChangesetBuilderTest.scala [hunk 4]
```diff
@@ -375,10 +375,78 @@ class ChangesetBuilderTest extends CompilerTest {
...
+          |""".stripMargin.linesIterator.mkString("\n")
+      val edit =
+        TextEdit(Range(Position(8, 88), Position(8, 88)), "Boolean.True")
```
Comment (4e6) [inline engine/runtime/src/test/scala/org/enso/compiler/test/context/ChangesetBuilderTest.scala]: You're editing `text3 = @` line (line 8), and overflowing on the next line by adding an extra `index`. While it works, it is very confusing

Comment (4e6) [reply engine/runtime/src/test/scala/org/enso/compiler/test/context/ChangesetBuilderTest.scala]: The correct edit is
```suggestion
        TextEdit(Range(Position(9, 72), Position(9, 72)), " Boolean.True")
```

File: engine/runtime/src/main/scala/org/enso/compiler/context/ChangesetBuilder.scala (hunk 1 above)
Comment (4e6) [reply engine/runtime/src/main/scala/org/enso/compiler/context/ChangesetBuilder.scala]: I have a hard time understanding how it works. Can you write it down as a comment on the function

File: engine/runtime/src/main/scala/org/enso/compiler/context/ChangesetBuilder.scala [hunk 5]
```diff
@@ -371,12 +403,18 @@ object ChangesetBuilder {
     * @param edit the location of the edit
     * @return the invalidated nodes of the tree
     */
-  private def invalidated(tree: Tree, edit: Location): Tree = {
+  private def invalidated(
+    tree: Tree,
+    edit: Location,
+    onlyLeafs: Boolean
```
Comment (4e6) [inline engine/runtime/src/main/scala/org/enso/compiler/context/ChangesetBuilder.scala:423]: and why do we need this parameter now?

File: engine/runtime/src/test/scala/org/enso/compiler/test/context/ChangesetBuilderTest.scala (hunk 4 above)
Comment (JaroslavTulach) [reply engine/runtime/src/test/scala/org/enso/compiler/test/context/ChangesetBuilderTest.scala]: I was wondering why I needed to increase the position to 88 when I saw just 72 in the debugger?

File: engine/runtime/src/test/scala/org/enso/compiler/test/context/ChangesetBuilderTest.scala (hunk 3 above)
Comment (JaroslavTulach) [reply engine/runtime/src/test/scala/org/enso/compiler/test/context/ChangesetBuilderTest.scala]: Removed in 8a2671b40a, thank you.

File: engine/runtime/src/main/scala/org/enso/compiler/context/ChangesetBuilder.scala (hunk 5 above)
Comment (JaroslavTulach) [reply engine/runtime/src/main/scala/org/enso/compiler/context/ChangesetBuilder.scala:423]: Without f867313 the _"should module with undefined literal"_ test fails:
```
$ git show f867313 | patch -p1 -R
$ sbt runtime/testOnly *ChangesetBuilderTest
[info] - should module with undefined literal *** FAILED ***
[info]   Set(b1c393b2-67be-488b-b46d-2adba21bca6d, c05a82f3-9dea-49fb-81ce-8dfa5703e5ac)
 did not contain the same elements as
 List(c05a82f3-9dea-49fb-81ce-8dfa5703e5ac) (ChangesetBuilderTest.scala:368)
```
The additional `b1c393b2-67be-488b-b46d-2adba21bca6d` id represents `1 + undefined` text. The `fillInHoles` function associates the space between `+` and `undefined` to `b1c393b2-67be-488b-b46d-2adba21bca6d` - as the edit replaces `undefined` with `42` it touches the space on the boundary and as such we get two change sets. To keep the original behavior we first work on _"leafs"_ only as before and only when they find nothing, we try harder with non-leaf elements.
```diff
--- engine/runtime/src/test/scala/org/enso/compiler/test/context/ChangesetBuilderTest.scala
+++ engine/runtime/src/test/scala/org/enso/compiler/test/context/ChangesetBuilderTest.scala
@@ -366,6 +366,7 @@ class ChangesetBuilderTest extends CompilerTest {
         .originalName
 
       invalidated(ir, code, edit) should contain theSameElementsAs Seq(
+        UUID.fromString("b1c393b2-67be-488b-b46d-2adba21bca6d"),
         undefinedName.getId
       )
       invalidatedAll(ir, code, edit) should contain theSameElementsAs Seq(
```
Alternative is to modify the test and ignore or accept the additional result.

**Comment by JaroslavTulach** [comment]
> ... understanding how it works. Can you write it down as a comment on the function

Using descriptive identifier names since 07fc59c

File: engine/runtime/src/main/scala/org/enso/compiler/context/ChangesetBuilder.scala (hunk 1 above)
Comment (JaroslavTulach) [reply engine/runtime/src/main/scala/org/enso/compiler/context/ChangesetBuilder.scala]: Using descriptive identifier names since 07fc59c

**Review by 4e6 — APPROVED** [review]

**Pull request merged**